File: go/src/csum/csum.go
// go build csum.go
// ./csum myfile.txt
// -- or --
// go run csum.go myfile.txt
// go run csum.go -- csum.go

// http://stackoverflow.com/questions/1821811/how-to-read-write-from-to-file

package main

import (
	// xxx not ";" sep since unused-import error ... we really need easy
	// import/not with "//".
	"csums"
	"flag"
	"fmt"
	"io"
	"log"
	"os"
	"strings"
)

// ----------------------------------------------------------------
func usage() {
	fmt.Fprintf(os.Stderr, "Usage: %s [options] {filenames ...}\n", os.Args[0])
	fmt.Fprintf(os.Stderr, "If no filenames are given, stdin is read.\n")
	flag.PrintDefaults()
	os.Exit(1)
}

// ----------------------------------------------------------------
func main() {
	// http://golang.org/pkg/flag
	pDoSpin   := flag.Bool("spin", false, "print running checksums to screen (default off)")
	// xxx string-set ... maybe do my own non-flag here?
	// xxx make a list-algos opt ...
	algoNames := strings.Join(csums.ChecksummerFactoryAlgoNames, ", ")
	pAlgo    := flag.String("algo", "crc64", "One of: " + algoNames + ".")

	flag.Usage = usage
	flag.Parse()
	args := flag.Args()
	doSpin := *pDoSpin

	checksummer, err := csums.ChecksummerFactory(*pAlgo)
	if (checksummer == nil) {
		log.Fatal(err)
	}

	ok := true
	if len(args) == 0 {
		ok = csum("-", checksummer, doSpin) && ok
	} else {
		for _, arg := range args {
			// ok && count(arg): not called after error
			ok = csum(arg, checksummer, doSpin) && ok
		}
	}
	if ok {
		os.Exit(0)
	} else {
		os.Exit(1)
	}
}

// ----------------------------------------------------------------
func csum(fileName string, checksummer csums.Checksummer, doSpin bool) (ok bool) {

	sourceStream := os.Stdin
	if fileName != "-" {
		var err error
		if sourceStream, err = os.Open(fileName); err != nil {
			log.Println(err)
			return false
		}
	}

	bufferSize := 2048

	buffer := make([]byte, bufferSize)
	eof := false
	var nblocks uint64 = 0
	var nbytes  uint64 = 0

	checksummer.Start()

	if (doSpin) {
		fmt.Print("...")
		os.Stdout.Sync()
	}

	for !eof {
		numBytesRead, err := sourceStream.Read(buffer)
		if err == io.EOF {
			err = nil
			eof = true
		} else if err != nil {
			log.Println(err)
			if fileName != "-" {
				sourceStream.Close()
			}
			return false
		} else {
			nblocks ++
			nbytes += uint64(numBytesRead)

			checksummer.Accumulate(buffer, numBytesRead)

			if (doSpin && ((nblocks & 0x3ff) == 0)) {
				fmt.Print("\r"+checksummer.GetStringState())
				os.Stdout.Sync()
			}
		}
	}

	if fileName != "-" {
		sourceStream.Close()
	}

	checksummer.Finish()
	stringSum := checksummer.GetStringSum()

	if (doSpin) {
		fmt.Print("\r")
	}
	fmt.Printf("%s  %6d  %s\n", stringSum, nbytes, fileName)

	return true
}

File: scala/csum/Csum.scala
// ================================================================
// Copyright (c) 2004, 2013 John Kerl.
// kerl at math dot arizona dot edu
//
// This code and information is provided as is without warranty of
// any kind, either expressed or implied, including but not limited to
// the implied warranties of merchantability and/or fitness for a
// particular purpose.
//
// No restrictions are placed on copy or reuse of this code, as long
// as these paragraphs are included in the code.
// ================================================================

// xxx scala.io.Source can't handle 8-bit binary data.
import java.io.BufferedInputStream
import java.io.FileInputStream
import java.io.InputStream

object Csum {

	// ----------------------------------------------------------------
	def usage() {
		System.err.println("Csum [options] {filenames ...}")
		System.err.println("If no filenames are given, stdin is read.")
		System.err.println("Options:")
		System.err.println("--eth:     Use Ethernet checksum algorithm (default)")
		System.err.println("--simple:  Use simple bytecount-bytesum checksum algorithm")
		System.err.println("--spin:    print running checksums to screen (default off)")
		System.err.println("--nospin:  do not print running checksums to screen")
		System.exit(1)
	}

	// ----------------------------------------------------------------
	// xxx scopt/scopt?
	def parseCommandLine(argv: Array[String])
		: Tuple2[Map[String,String],List[String]] =
	{
		var opts = Map("spin" -> "false", "algo" -> "eth")
		var args = List[String]()

		for (arg <- argv) {
			// xxx case?
			if (arg.startsWith("-")) {
				if (arg == "--spin")
					opts += "spin" -> "true"
				else if (arg == "--nospin")
					opts += "spin" -> "false"
				else if (arg == "--ip")
					opts += "algo" -> "ip"
				else if (arg == "--eth")
					opts += "algo" -> "eth"
				else if (arg == "--simple")
					opts += "algo" -> "simple"
				else
					usage()
			}
			else {
				args ::= arg
			}
		}

		(opts, args)
	}

	// ----------------------------------------------------------------
	def main(argv: Array[String]) {
		val (opts, args) = parseCommandLine(argv)

		if (args.length == 0)
			doOneArgument("-", opts)
		else
			args.foreach(arg => doOneArgument(arg, opts))
	}

	// ----------------------------------------------------------------
	def doOneArgument(fileName: String, opts: Map[String,String]) {
		val bufferSize = 1024

		val inputStream = if (fileName == "-") System.in else new FileInputStream(fileName)
		val bis     = new BufferedInputStream(inputStream)
		val buffer  = new Array[Byte](bufferSize)
		val doSpin  = opts("spin") == "true"
		var done    = false
		var nblocks = 0
		var nbytes  = 0

		val checksummer = ChecksummerFactory(opts("algo"))

		checksummer.start

		if (doSpin) {
			print("...")
			System.out.flush
		}

		while (!done) {
			val numBytesRead = bis.read(buffer, 0, bufferSize)
			if (numBytesRead < 0) {
				done = true
			}
			else {
				nblocks += 1
				nbytes += numBytesRead
				checksummer.accumulate(buffer, numBytesRead)

				if (doSpin && ((nblocks & 0xfff) == 0)) {
					print("\r"+checksummer.getStringState)
					System.out.flush
				}
			}
		}
		checksummer.finish

		val stringSum = checksummer.getStringSum
		if (doSpin)
			print("\r")
		println(stringSum + "  " + nbytes + "  " + fileName)
	}
}
